cmake/subtree-leveldb.cmake

```diff
@@ -0,0 +1,111 @@
+# Subtree at https://github.com/bitcoin-core/leveldb-subtree
+# Upstream at https://github.com/google/leveldb
+add_library(leveldb STATIC EXCLUDE_FROM_ALL
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/builder.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/c.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/dbformat.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/db_impl.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/db_iter.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/dumpfile.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/filename.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/log_reader.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/log_writer.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/memtable.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/repair.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/table_cache.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/version_edit.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/version_set.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/db/write_batch.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/helpers/memenv/memenv.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/table/block_builder.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/table/block.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/table/filter_block.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/table/format.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/table/iterator.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/table/merger.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/table/table_builder.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/table/table.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/table/two_level_iterator.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/arena.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/bloom.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/cache.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/coding.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/comparator.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/crc32c.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/env.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/filter_policy.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/hash.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/histogram.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/logging.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/options.cc
+  ${CMAKE_SOURCE_DIR}/src/leveldb/util/status.cc
+)
+if(WIN32)
+  target_sources(leveldb
+    PRIVATE
+      ${CMAKE_SOURCE_DIR}/src/leveldb/util/env_windows.cc
+  )
+else()
+  target_sources(leveldb
+    PRIVATE
+      ${CMAKE_SOURCE_DIR}/src/leveldb/util/env_posix.cc
+  )
+endif()
+
+target_include_directories(leveldb
+  PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/src/leveldb>
+    $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/src/leveldb/helpers/memenv>
+    $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/src/leveldb/include>
+)
+
+target_compile_definitions(leveldb
+  PRIVATE
+    __STDC_LIMIT_MACROS
+    HAVE_SNAPPY=0
+    HAVE_CRC32C=1
+    FALLTHROUGH_INTENDED=[[fallthrough]]
+    LEVELDB_IS_BIG_ENDIAN=${WORDS_BIGENDIAN}
+)
+if(HAVE_FDATASYNC)
+  target_compile_definitions(leveldb PRIVATE HAVE_FDATASYNC=1)
+else()
+  target_compile_definitions(leveldb PRIVATE HAVE_FDATASYNC=0)
```

Comment:
Do we actually want this? LevelDB's own CMake stuff acts on `HAVE_FDATASYNC` being *defined* rather than 1...

Reply:
> LevelDB's own CMake stuff acts on `HAVE_FDATASYNC` being _defined_ rather than 1...

I don't think so.[ "LevelDB's own CMake stuff"](https://github.com/bitcoin-core/leveldb-subtree/blob/330dd6235f50c01e49b4cfadf6df9e026dc61f0d/port/port_config.h.in#L8-L11):
```
// Define to 1 if you have a definition for fdatasync() in <unistd.h>.
#if !defined(HAVE_FDATASYNC)
#cmakedefine01 HAVE_FDATASYNC
#endif  // !defined(HAVE_FDATASYNC)
```
_always_ defines the `HAVE_FDATASYNC` macro via [`#cmakedefine01`](https://cmake.org/cmake/help/latest/command/configure_file.html) replacement.